mkdocs/docs/meili/v0.0.3/orchestration_thoughts.md

```diff
@@ -0,0 +1,55 @@
+
+---
+
+With these design philosophies, we can go ahead and actually plan out how we want to handle npm orchestration.
+
+## Commands and Architecture
+
+The first thing I want to cover is commands. Ideally we run `cloesce run` and everything is generated, all the user has to do is `wrangler run dev`
+
+But what goes under the hood? First thing I want to make clear is that npm should not just be a underlying bash script. We should treat the build as a dataflow, not a bunch of shell commands. **Essentially we need to build a directed acyclic graph since most good npm packages use frameworks or have their own systems that do this.** Each step (TS to IR, IR to workers.ts, IR to SQL, Wrangler build, etc.) is a task node which consumes inputs and produces outputs. So the edges say "the output of this task node will feed into the next task", and when we rebuild we can detect what the user changed, find only the tasks downstream that need to be rerun, and rebuild the minimal set. This means that things will be way faster when we rebuild and we have observability into what changed for logging purposes.
+
+![Acrylic Graph](../../assets/orchestration.png)
+
+## Implementation
+
+> **Note:** There's **tooling frameworks that might just do all this for us like Nx** so we don't have to deal with all this set up and caching that I might just go with for the sake of simplicity. This just describes iff we want to go down the path of building from the ground up for whatever reason. If the frameworks that are avaliable aren't enough and we need to build this from the ground up, I can see pretty easily getting bogged down here, so as a first goal I want to just first just get everything to run in order correctly. Then I'll move onto implementing the DAG so we can run everything in parallel and reduce runtime (It might seem unnecessary to set up DAG here since there's easier ways to get the same result, but the DAG is critical for caching and faster rebuilds moving forward), then move onto caching builds for quick rebuilds.
+
+Implementation wise, we need to make sure each run behaves the same and reads one source of truth, every file should be generated into the same `.generated/` file. Then we need to do content hashing and env fingerprints, meaning we read the bytes of important files like the TS models and hash them. This ensures that when we rebuild and the hash is the same, we can quickly keep going and build since we know nothing has changed. Then we formalize each build step as a worker with a clear contract, as in each step should have dependencies that must be completed before we run anything, what counts as inputs, the output it promises to write, and how we run everything. Then we build the DAG and do topological sort, we build the dependency graph from the registry and compute a safe execution order, skipping cache hits and executing everything in one go.
```

Comment:
for the sake of this PR we should stay simple: save a hash, if anything is changed at all just re-run the entire generate command. no dependencies, same execution order as defined in e2e tests, always ran the same way